docs/glosary.rst

```diff
@@ -0,0 +1,72 @@
+Glossary of Low-level Concepts in Mathics Design
+================================================
+
+
+Verbs
+-----
+
+* Rewrite: is an operation over an Expression that produces a new Expression by reorganizing the head and the leaves of an expression and its subexpressions without information of the ``Definition`` s associated with the expression. Operations like ``Flatten``, ``Thread``, ``Sort``, ``Apply``, ``Part`` and ``Drop`` in WL are examples of this kind of operation.
```

Comment:
> `Definition`` s associated with the expression.

@mmatera Doesn't the rewrite process consult the definition attributes in checking to see if a rule is applicable? 

For example in
 ```
"StringReverse[string_String]"
```

how does the rule matching figure out whether something is a string?  

Is the "Orderless" property used in rewriting, and if so does that come from the Definition?

Reply:
I have a somewhat strong dislike for either `evaluation_step()` and or `sort_evaluate()` so I think it best to shelve the function name renaming and focus more on describing what is going on and what each function does in a docstring. 

I probably wouldn't have objected to `evaluate_next()` if it had been clearly described somewhere. And I suspect if it had been clearly described, the author would have realized the lameness, ambiguity and vagueness of the name.

Reply:
Regarding *rewrite* I would be OK with using the Wikipedia definition, but it does not fit with the mathics design. Maybe, this is because an issue in the current design, or maybe is because the concept does not fit with the wl design.

I think that to be able to explain something we need to understand very well the subject. Then, as we gain clarity we are going to be in a better position to write a better documentation.

Reply:
> Regarding _rewrite_ I would be OK with using the Wikipedia definition, but it does not fit with the mathics design. 

Ok - let's check this: 

> In mathematics, computer science, and logic, 

Yes, we are talking about mathematics and computer science here. Possibly not logic but "and" in the English language is commonly used as a logical "or". 

> rewriting covers a wide range of methods of replacing subterms of a formula with other terms. 

Yes, that fits here too. Our "formulae" are expressed as an expression. 

> Such methods may be achieved by rewriting systems (also known as rewrite systems, rewrite engines or reduction systems). In their most basic form, they consist of a set of objects, plus relations on how to transform those objects.

That seems to fit here as, we "objects" here refers to expressions, a rule is a kind of relation which specifies how to transform an "object" or expression into another "object". 

> Rewriting can be non-deterministic. One rule to rewrite a term could be applied in many different ways to that term, or more than one rule could be applicable. 

Yes, that fits here as well. Mathics/WL rules can be applied in many different way and more than one rule can be applicable. 

> Rewriting systems then do not provide an  for changing one term to another, but a set of possible rule applications. When combined with an appropriate algorithm, however, rewrite systems can be viewed as computer programs. 

And this is true here. It is possible to create a set of rules the works purely on symbols (that is having no value other than the name of the symbol) and with no eval portion (e.g computation for what a specific operator like Plus does).

And this is no coincidence. Wolfram was familiar with rewriting systems and based Mathematica on its principles. 

I get that you want to describe something specific or at a different level here or how Mathics currently works to some specific level of detail. There is time to do this and a place in the doc for when we get around to that level of detail.  But in  doing so we should not preempt the normal definitions of things.